SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestShutdownPreparation/SpatialTestShutdownPreparationTrigger.cpp

```diff
@@ -0,0 +1,143 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#include "SpatialTestShutdownPreparationTrigger.h"
+#include "EngineClasses/SpatialNetDriver.h"
+#include "SpatialFunctionalTestFlowController.h"
+#include "SpatialFunctionalTestLBDelegationInterface.h"
+#include "SpatialGDKServicesConstants.h"
+#include "SpatialGDKServicesModule.h"
+#include "TestPrepareShutdownListener.h"
+
+ASpatialTestShutdownPreparationTrigger::ASpatialTestShutdownPreparationTrigger()
+{
+	Author = "Tilman Schmidt";
+	Description = TEXT("Trigger shutdown preparation via worker flags and make sure callbacks get called in C++ and Blueprints");
+	StepTimer = 0.0f;
+	LocalListener = nullptr;
+}
+
+void ASpatialTestShutdownPreparationTrigger::BeginPlay()
+{
+	Super::BeginPlay();
+	{ // Step 1 - Test print on all workers
+		AddStep(TEXT("AllWorkers_SetupListener"), FWorkerDefinition::AllWorkers, nullptr, [this]() {
+			UWorld* World = GetWorld();
+
+			// Spawn a non-replicated actor that will listen for the shutdown event.
+			// Using a non-replicated actor since this is the easiest way to make sure that every worker has exactly one instance of it.
+			LocalListener =
+				World->SpawnActor<ATestPrepareShutdownListener>(PrepareShutdownListenerClass, FVector::ZeroVector, FRotator::ZeroRotator);
+			AssertTrue(IsValid(LocalListener), TEXT("Listener actor is valid."));
+			RegisterAutoDestroyActor(LocalListener);
+
+			LocalListener->RegisterCallback();
+			if (LocalListener->NativePrepareShutdownEventCount != 0 || LocalListener->BlueprintPrepareShutdownEventCount != 0)
+			{
+				UE_LOG(LogTemp, Log, TEXT("Failing test due to event counts starting out wrong. native: %d, blueprint: %d"),
+					   LocalListener->NativePrepareShutdownEventCount, LocalListener->BlueprintPrepareShutdownEventCount);
+				FinishTest(EFunctionalTestResult::Failed, TEXT("Number of triggered events should start out at 0"));
+				return;
+			}
+
+			FinishStep();
+		});
+
+		AddStep(TEXT("Server1_TriggerShutdownPreparation1"), FWorkerDefinition::Server(1), nullptr, [this]() {
+			FString WorkerFlagSetArgs = TEXT("local worker-flag set UnrealWorker PrepareShutdown Yes --local_service_grpc_port 9876");
+
+			FString WorkerFlagSetResult;
+			FString StdErr;
+			int32 ExitCode;
+			FPlatformProcess::ExecProcess(*SpatialGDKServicesConstants::SpatialExe, *WorkerFlagSetArgs, &ExitCode, &WorkerFlagSetResult,
+										  &StdErr, *SpatialGDKServicesConstants::SpatialOSDirectory);
+
+			if (ExitCode != 0)
+			{
+				FinishTest(EFunctionalTestResult::Error, TEXT("Setting the worker flag failed"));
+				return;
+			}
+
+			FinishStep();
+		});
+
+		AddStep(TEXT("AllServers_CheckEventHasTriggered"), FWorkerDefinition::AllServers, nullptr, nullptr, [this](float DeltaTime) {
```

Comment:
What's the default timeout here?

Reply:
From what I can tell, individual steps have no timeout by default, but the overall test has a 1 minute time limit.